File: frontend/packages/checksums/src/schema.ts
import type { PluginSchema } from '@sep/api';

/**
 * Checksums plugin schema — defines the form and list view for pt-table-checksum tasks.
 *
 * This is the ONLY plugin-specific configuration needed.
 * The SchemaDrivenPlugin component auto-generates list, create, and detail pages
 * entirely from this definition.
 *
 * In production, this schema will be served by the backend at
 * GET /api/plugins/checksums/schema — this mock is used for development.
 */
export const checksumsSchema: PluginSchema = {
  name: 'checksums',
  displayName: 'Checksums',
  description: 'Run pt-table-checksum to verify data consistency between MySQL replicas.',
  taskType: 'pt-table-checksum',

  forms: [
    {
      title: 'Target',
      description: 'Select the MySQL service to run checksums against.',
      fields: [
        {
          name: 'serviceId',
          label: 'MySQL Service',
          type: 'service',
          required: true,
          serviceTypes: ['mysql'],
        },
        {
          name: 'schema',
          label: 'Schema',
          type: 'string',
          placeholder: 'Leave empty for all schemas',
          description: 'Optionally filter to a specific database schema.',
        },
      ],
    },
    {
      title: 'Options',
      fields: [
        {
          name: 'chunkSize',
          label: 'Chunk Size',
          type: 'choice',
          default: '1000',
          choices: [
            { label: '1,000', value: '1000' },
            { label: '5,000', value: '5000' },
            { label: '10,000', value: '10000' },
            { label: '50,000', value: '50000' },
          ],
        },
        {
          name: 'replicateCheck',
          label: 'Replicate check (--replicate-check)',
          type: 'bool',
          default: true,
        },
        {
          name: 'checkInterval',
          label: 'Check interval (seconds)',
          type: 'integer',
          default: 1,
          ge: 1,
          le: 3600,
        },
      ],
    },
  ],

  capabilities: {
    alertOnFail: true,
    scheduling: true,
  },

  listView: {
    columns: [
      { key: 'id', label: 'ID', sortable: true },
      { key: 'service', label: 'Service', format: 'chip' },
      { key: 'schema', label: 'Schema', format: 'code' },
      { key: 'status', label: 'Status', format: 'status', sortable: true },
      { key: 'differences', label: 'Differences' },
      { key: 'lastRun', label: 'Last Run', format: 'relative', sortable: true },
    ],
    defaultSort: 'lastRun',
  },
};

File: app/sep/plugins/checksums/schema.py
# Copyright (C) 2026 Percona LLC
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.

"""Define the PluginSchema for the Checksums plugin."""

from app.inventory.models import ServiceTypeEnum
from app.sep.plugins.framework.schema import (
    BoolField,
    Capabilities,
    Choice,
    ChoiceField,
    Column,
    ColumnFormat,
    FormSection,
    HostField,
    ListView,
    PluginSchema,
    ServiceField,
    StringField,
)

checksums_schema = PluginSchema(
    name="checksums",
    display_name="Checksums",
    description="Run pt-table-checksum to verify MySQL replication consistency.",
    forms=[
        FormSection(
            title="Task",
            fields=[
                StringField(
                    name="task_name",
                    label="Task Name",
                    required=True,
                ),
                HostField(
                    name="hostname",
                    label="Executor Host",
                    required=True,
                ),
                ServiceField(
                    name="service_id",
                    label="Database Host",
                    required=True,
                    service_types=[ServiceTypeEnum.MYSQL],
                ),
            ],
        ),
        FormSection(
            title="Data",
            fields=[
                StringField(
                    name="databases",
                    label="Databases",
                    description="Comma-separated database names",
                ),
                StringField(
                    name="tables",
                    label="Tables",
                    description="Comma-separated table names (schema.table format)",
                ),
            ],
        ),
        FormSection(
            title="Recursion",
            fields=[
                ChoiceField(
                    name="recursion_method",
                    label="Recursion Method",
                    required=True,
                    default="processlist",
                    choices=[
                        Choice(label="Default", value="default"),
                        Choice(label="Processlist", value="processlist"),
                        Choice(label="Hosts", value="hosts"),
                        Choice(label="DSN", value="dsn"),
                        Choice(label="None", value="none"),
                    ],
                ),
                StringField(
                    name="dsn_table",
                    label="DSN Table",
                    default="D=percona,t=dsns",
                    description="Only used when recursion method is 'dsn'",
                ),
            ],
        ),
        FormSection(
            title="Flags",
            fields=[
                BoolField(
                    name="binary_index",
                    label="Binary Index",
                    description="Use BLOB type for replicate-table boundary columns",
                    default=False,
                ),
                BoolField(
                    name="explain_arg",
                    label="Explain (dry run)",
                    description="Show but do not execute checksum queries",
                    default=False,
                ),
                BoolField(
                    name="fail_on_stopped_replication",
                    label="Fail on Stopped Replication",
                    description="Fail with an error if replication is stopped",
                    default=False,
                ),
                BoolField(
                    name="truncate_replicate_table",
                    label="Truncate Replicate Table",
                    description="Truncate the replicate table before starting",
                    default=False,
                ),
            ],
        ),
        FormSection(
            title="Advanced",
            fields=[
                StringField(
                    name="pause_file",
                    label="Pause File",
                    description="Execution pauses while this file exists",
                ),
                StringField(
                    name="progress",
                    label="Progress",
                    default="time,10",
                    description="Print progress reports to STDERR (e.g. time,10)",
                ),
                StringField(
                    name="set_vars",
                    label="Set Vars",
                    default="transaction_isolation='READ-COMMITTED',lock_wait_timeout=5",
                    description="MySQL variables to set (comma-separated key=value pairs)",
                ),
                StringField(
                    name="max_load",
                    label="Max Load",
                    default="Threads_running=50",
                    description="Pause when any GLOBAL STATUS variable exceeds this threshold",
                ),
                StringField(
                    name="chunk_time",
                    label="Chunk Time",
                    default="0.5",
                    description="Target execution time per chunk in seconds",
                ),
                StringField(
                    name="max_lag",
                    label="Max Lag",
                    default="150",
                    description="Pause until replica lag falls below this value (seconds)",
                ),
            ],
        ),
    ],
    capabilities=Capabilities(
        chaining=True, alert_on_fail=True, scheduling=True, stats=True
    ),
    list_view=ListView(
        columns=[
            Column(key="name", label="Name", sortable=True),
            Column(key="status", label="Status", format=ColumnFormat.STATUS),
            Column(key="service_type", label="Service Type", format=ColumnFormat.CHIP),
            Column(key="created_at", label="Created", format=ColumnFormat.RELATIVE),
            Column(key="created_by", label="Created By"),
        ],
    ),
)
